Add support for go.delivery.report.fields

kevinconaway opened the pull request:
@edenhill 
This configuration property will allow users to enable/disable certain fields of the delivery reports to reduce allocations.

Only the key & value are currently supported

Addresses #524

kevinconaway commented:
@edenhill any update here ?

edenhill commented on kafka/handle.go:
```diff
@@ -113,6 +113,9 @@ type handle struct {
 	fwdDr bool
 
+	// Enabled fields for delivery reports
+	mf *messageFields
```
Can we have a more descriptive field name than mf? msgFields?

edenhill commented on kafka/producer.go:
```diff
@@ -423,6 +424,8 @@ func (p *Producer) Purge(flags int) error {
 //   go.delivery.reports (bool, true) - Forward per-message delivery reports to the
 //                                      Events() channel.
+//   go.delivery.report.fields (string, all) - Comma separated list of fields to enable for delivery reports.
+//											   Allowed values: all, none, key, value
```
Use space for indent

edenhill commented on kafka/producer.go:
```diff
@@ -461,6 +464,27 @@ func NewProducer(conf *ConfigMap) (*Producer, error) {
+		return nil, err
+	}
+	p.handle.mf = newMessageFields()
+	switch v {
```
I think it would be better to move the following logic into newMessageFields

edenhill commented on kafka/producer.go:
```diff
@@ -461,6 +464,27 @@ func NewProducer(conf *ConfigMap) (*Producer, error) {
+				p.handle.mf.key = true
+			case "value":
+				p.handle.mf.value = true
+			}
```
Should error out on unknown fields.

edenhill commented on kafka/handle.go:
```diff
@@ -323,3 +329,24 @@ func (h *handle) setOAuthBearerTokenFailure(errstr string) error {
 	return newError(cErr)
 }
+
+// messageFields controls which fields are deserialized for producer delivery reports & incoming messages
```
They're not really deserialized, maybe "which fields are made available"?

edenhill commented on kafka/handle.go:
```diff
@@ -323,3 +329,24 @@ func (h *handle) setOAuthBearerTokenFailure(errstr string) error {
+// messageFields controls which fields are deserialized for producer delivery reports & incoming messages
+// true values indicate that the field should be included
+type messageFields struct {
+	key   bool
```
` // Message.Key` and `// Message.Value`

edenhill reviewed (changes requested):
Left some comments.
Please add a test case producer_test.go where you produce a message with a short message.timeout.ms (1ms) and verify that only the enabled messageFields are available.

kevinconaway commented:
@edenhill Sorry about the delay here.  

I believe that I have addressed your comments, please let me know what, if anything, else needs doing.  Thank you!

kevinconaway commented:
@edenhill thoughts here?

edenhill commented:
Looks good!
Can you rebase the PR on latest master to fix the conflicts?

kevinconaway commented:
Done although I didn't see any merge conflicts here.  Is there something I'm not seeing?

edenhill commented:
Huhm, github says this:
> This branch cannot be rebased due to conflicts
> Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.

Try to rebase and squash your PR on top of master instead of merging master, like so:
```bash
$ git fetch upstream   # or whatever our repo remote is named)
$ git rebase -i origin/upstream 
# In the editor, fixup (f)  all commits into one single commit. Save and quit.
$ git push --force ......
```

kevinconaway commented:
Done.  The PR dialog still doesn't show any conflicts for me, how strange

<img width="1006" alt="Screen Shot 2021-02-17 at 8 47 45 AM" src="https://user-images.githubusercontent.com/6921953/108213389-f803cc00-70fc-11eb-8f28-ba9af544b2aa.png">

edenhill commented:
Looks good now, thanks!

edenhill commented:
[clabot:check]

edenhill commented:
Thank you, Kevin!
